motion/thruster_interface_auv_cpp/src/thruster_interface_auv_driver.cpp

```diff
@@ -0,0 +1,138 @@
+    }
+
+    // Select the appropriate coefficients based on the operational voltage
+    auto left_coeffs = COEFFS[SYSTEM_OPERATIONAL_VOLTAGE]["LEFT"];
+    auto right_coeffs = COEFFS[SYSTEM_OPERATIONAL_VOLTAGE]["RIGHT"];
+
+    std::vector<int16_t> interpolated_pwm;
+    for (size_t i = 0; i < forces_in_kg.size(); ++i) {
+        double force = forces_in_kg[i];
+        double pwm = 0.0;
+        if (force < 0) {
+            pwm =   left_coeffs[0] * std::pow(forces_in_kg[i], 3) + \
+                    left_coeffs[1] * std::pow(forces_in_kg[i], 2) + \
+                    left_coeffs[2] * forces_in_kg[i] + \
+                    left_coeffs[3];
+        } else if (force == 0.0) {
+            pwm = 1500;
+        } else {
+            pwm =   right_coeffs[0] * std::pow(forces_in_kg[i], 3) + \
+                    right_coeffs[1] * std::pow(forces_in_kg[i], 2) + \
+                    right_coeffs[2] * forces_in_kg[i] + \
+                    right_coeffs[3];
+        }
+        interpolated_pwm.push_back(static_cast<int16_t>(pwm));
+    }
+
+    return interpolated_pwm;
+}
+
+void ThrusterInterfaceAUVDriver::send_data_to_escs(const std::vector<int16_t>& thruster_pwm_array) {
+    uint8_t i2c_data_array[16];
+    for (size_t i = 0; i < thruster_pwm_array.size(); ++i) {
+        i2c_data_array[2*i] = (thruster_pwm_array[i] >> 8) & 0xFF; // MSB
+        i2c_data_array[2*i + 1] = thruster_pwm_array[i] & 0xFF;    // LSB
+    }
+
+    // Set the I2C slave address
+    if (ioctl(bus_fd, I2C_SLAVE, PICO_I2C_ADDRESS) < 0) {
+        std::cerr << "ERROR: Failed to set I2C slave address" << std::endl;
+        return;
+    }
+
+    // Write data to the I2C device
+    if (write(bus_fd, i2c_data_array, 16) != 16) {
+        std::cerr << "ERROR: Failed to write to I2C device" << std::endl;
+    }
+}
+
+std::vector<int16_t> ThrusterInterfaceAUVDriver::drive_thrusters(const std::vector<double>& thruster_forces_array) {
+    // Apply thruster mapping and direction
+    std::vector<double> mapped_forces(thruster_forces_array.size());
+    for (size_t i = 0; i < THRUSTER_MAPPING.size(); ++i) {
+        mapped_forces[i] = thruster_forces_array[THRUSTER_MAPPING[i]] * THRUSTER_DIRECTION[i];
+    }
+
+    // Convert forces to PWM
+    std::vector<int16_t> thruster_pwm_array = interpolate_forces_to_pwm(mapped_forces);
+
+    // Apply thruster offset and limit PWM if needed
+    for (size_t i = 0; i < thruster_pwm_array.size(); ++i) {
+        //thruster_pwm_array[i] += THRUSTER_PWM_OFFSET[i];
+
+        // Clamp the PWM signal
+        if (thruster_pwm_array[i] < PWM_MIN[i]) {
+            thruster_pwm_array[i] = PWM_MIN[i];
+        } else if (thruster_pwm_array[i] > PWM_MAX[i]) {
+            thruster_pwm_array[i] = PWM_MAX[i];
+        }
+    }
+
+    // Send data through I2C
+    try {
+        send_data_to_escs(thruster_pwm_array);
+    } catch (...) {
+        std::cerr << "ERROR: Failed to send PWM values" << std::endl;
+    }
```

Comment:
cppcoreguidelines: [E.17: Don’t try to catch every exception in every function](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e17-dont-try-to-catch-every-exception-in-every-function)

If `send_data_to_escs` can throw a million different exceptions all with separate classes, it would be fine to leave it as-is, but if it has a reasonable amount of exceptions it may throw, catch them explicitly. Also, make sure to print the exception message, so you know what went wrong :smile: